Fix automatic removal of ingress sandbox when last service leaves

ctelfer opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/moby/moby/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**
Patched libnetwork to avoid automatically deleting ingress networks when the only endpoint remaining in the network was the ingress load balancer IP.  Then adjust the moby code so that it explicitly removes this endpoint when explicit ingress removal is requested.  The issue this fixes is detailed in https://github.com/docker/libnetwork/issues/2097.

**- How I did it**
Two separate patches are required.  The first is to libnetwork to avoid an issue introduced in https://github.com/docker/libnetwork/pull/2011 that automatically removes *any* network (including ingress) when the number of endpoints drops to 1 and the network contains a load balancing IP.  The patch prevents the removal if the network is marked as ingress.   The second patch adds logic to the daemon.releaseIngress() call in moby to remove the ingress load balancing endpoint before invoking libnetwork.network.Delete().

**- How to verify it**
I added a test to the moby integration tests called TestServiceWithIngressNetwork.   This test:
  *  spins up a swarm
  * waits for the ingress network to "settle"
  * creates a simple service and waits for it to converge
  * removes the service
  * waits and then verifies that the ingress network is still present and is not corrupted

If one applies the test patch (the first in the series) without the other two fix patches, the test will fail noting ingress corruption.  If one applies the entire PR set of patches, and runs the test, it will pass.  (running the test as `TESTFLAGS='-test.run TestServiceWithIngressNetwork' make test-integration`)

**- Description for the changelog**
Prevent implicit removal of the ingress network

**- A picture of a cute animal (not mandatory but encouraged)**

ctelfer committed:
Delete the load balancer endpoint in Ingress nets

Ingress networks will no longer automatically remove their
load-balancing endpoint (and sandbox) automatically when the network is
otherwise upopulated.   This is to prevent automatic removal of the
ingress networks when all the containers leave them.  Therefore
explicit removal of an ingress network also requires explicit removal
of its load-balancing endpoint.

Signed-off-by: Chris Telfer <ctelfer@docker.com>

ctelfer committed:
Add test for ingress removal on service removal

The commit https://github.com/moby/moby/pull/35422 had the result of
accidentally causing the removal of the ingress network when the
last member of a service left the network.  This did not appear
in swarm instances because the swarm manager would still maintain
and return cluster state about the network even though it had
removed its sandbox and endpoint.  This test verifies that after a
service gets added and removed that the ingress sandbox remains
in a functional state.

Signed-off-by: Chris Telfer <ctelfer@docker.com>

mavenugo reviewed (approved):
LGTM

mavenugo commented:
@ctelfer thanks for also adding the integration-cli test for this tricky test scenario.

ddebroy reviewed (approved):
LGTM

fcrisciani reviewed (approved):
LGTM

selansen commented:
LGTM

vdemeester reviewed (changes requested):
@ctelfer SGTM, but can you reference the upstream PR on `libnetwork` (we'll need to wait for it to get merged to be able to merge it here)

ctelfer commented:
@vdemeester Thanks for the review.  At the time I filed the PR, the libnetwork change wasn't merged.  I've updated the vendoring now so it references the appropriate libnetwork commit directly.  CI seems to be in progress.  For reference, the libnetwork PR for this fix is https://github.com/docker/libnetwork/issues/2097.  Thanks again!

thaJeztah commented:
@ctelfer looks like something went wrong with vendoring;

```
15:32:37 The result of vndr differs
15:32:37 
15:32:37  M vendor/github.com/docker/libnetwork/config/config.go
15:32:37  M vendor/github.com/docker/libnetwork/controller.go
15:32:37  M vendor/github.com/docker/libnetwork/drivers_ipam.go
15:32:37  M vendor/github.com/docker/libnetwork/ipams/builtin/builtin_unix.go
15:32:37  M vendor/github.com/docker/libnetwork/ipams/builtin/builtin_windows.go
15:32:37  M vendor/github.com/docker/libnetwork/ipamutils/utils.go
15:32:37  M vendor/github.com/docker/libnetwork/network.go
15:32:37  M vendor/github.com/docker/libnetwork/service_common.go
15:32:37  M vendor/github.com/docker/libnetwork/vendor.conf
15:32:37 
15:32:37 Please vendor your package with github.com/LK4D4/vndr.
```

ctelfer committed:
Update vendoring for libnetwork PR #2097

This PR prevents automatic removal of the load balancing sandbox
endpoint when the endpoint is the last one in the network but
the network is marked as ingress.

Signed-off-by: Chris Telfer <ctelfer@docker.com>

ctelfer commented:
*facepalm*  Updated the vendor.conf but forgot to actually run vndr.   :(

ctelfer commented:
Looks like a test in the experimental tests failed.  It appears to be an error in swarm teardown in TestAPISwarmRestartCluster.  Is there a way to easily run just this test so I can try to reproduce it?

vdemeester reviewed (approved):
LGTM :tiger2:

thaJeztah commented:
Full diff of libnetwork changes being brought in: https://github.com/docker/libnetwork/compare/ed2130d117c11c542327b4d5216a5db36770bc65...3aca383eb555510f3f17696f9505f7bfbd25f0e5

Which includes:

- https://github.com/docker/libnetwork/pull/2058 Allow user to specify default address pools for docker networks
  - libnetwork changes related to https://github.com/moby/moby/pull/36396
- https://github.com/docker/libnetwork/pull/2085 Add warning message for the failure of deleting link device
- https://github.com/docker/libnetwork/pull/2088 Limit max length transient log
- https://github.com/docker/libnetwork/pull/2094 fix for #1333, calling LinkDel to delete link device when the err is NULL
- https://github.com/docker/libnetwork/pull/2096 Fix crash if the len was < maxSetStringLen
- https://github.com/docker/libnetwork/pull/2101 Prevent ingress deletion when endpoint count == 1
  - (the changes related to this PR)
- https://github.com/docker/libnetwork/pull/2104 IPAM tests
- https://github.com/docker/libnetwork/pull/2105 Fix for Duplicate IP issues

thaJeztah commented on integration/network/service_test.go:
```diff
@@ -36,9 +36,67 @@ func TestServiceWithPredefinedNetwork(t *testing.T) {
+	poll.WaitOn(t, swarmIngressReady(client), pollSettings)
+
+	var instances uint64 = 1
+	serviceName := "TestIngressService"
```
We should try to get rid of names where possible (at least be sure they're unique).

I see there's a couple of other locations already, so also for a follow-up

thaJeztah replied:
Opened https://github.com/moby/moby/issues/36583 for that ^^

thaJeztah commented on integration/network/service_test.go:
```diff
@@ -36,9 +36,67 @@ func TestServiceWithPredefinedNetwork(t *testing.T) {
+
+}
+
+const ingressNet = "ingress"
```
Do we also need a test with a custom ingress network?

ctelfer replied:
That's a fair question.  I'm not aware of any code that tags any other networks besides a swarm-mode network named "ingress" as an ingress network.  Neither am I aware of our support for such a construct.  However, that may just be my lack of experience w/ the product.

My use of the named constant was more out of a general purpose aversion to magic constants sprinkled through code.  (i.e. collecting them to a named token makes refactoring easier and saner as a rule ... but will probably not be needed in this case)

ddebroy replied:
Regarding "any code that tags any network besides swarm-mode network named "ingress" as an ingress network": you can create an overlay network named anything after removing existing ingress as long as it marked with `--ingress` flag. E.g. (after deleting existing ingress) `docker network create --ingress -d overlay testingress` works

thaJeztah reviewed (approved):
SGTM, but left some questions/notes for possible follow-ups

ctelfer commented on daemon/network.go:
```diff
@@ -222,6 +222,8 @@ func (daemon *Daemon) releaseIngress(id string) {
 		return
 	}
 
+	daemon.deleteLoadBalancerSandbox(n)
```
Very fair question and one discussed a bit in https://github.com/docker/libnetwork/issues/2097.   Ultimately, this function was copied into libnetwork about 6 months ago but the original never got removed.   As it was, this was dead code.  There is a function of the same name and behavior in libnetwork and it is not exported.   It would be possible to just export said function (changing the libnetwork API, but in a small way).  However, given that 
    1. I'm working on a patch in the near term to refactor of load balancing to improve scalability
    2. This refactoring might affect how we expose said function
    3. We are trying to get this regression fixed for the immanent release
It seemed more prudent to make the smaller change to fix the regression and then clean up the duplicate code in the larger refactorization.

thaJeztah replied:
Thanks; yes it's not new code, so I'm good with keeping it for a follow-up. Just was looking at it, and it felt like a _lot_ of error conditions were ignored, so left that comment for later investigation if needed. 👍

ctelfer commented:
Thanks for all the help!

thaJeztah commented on integration/network/service_test.go:
```diff
@@ -36,9 +36,67 @@ func TestServiceWithPredefinedNetwork(t *testing.T) {
+	err = client.ServiceRemove(context.Background(), serviceID)
+	require.NoError(t, err)
+
+	poll.WaitOn(t, serviceIsRemoved(client, serviceID), pollSettings)
```
Opened https://github.com/moby/moby/issues/36581 for that
